astropy/units/structured.py

```diff
@@ -0,0 +1,480 @@
+# -*- coding: utf-8 -*-
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+This module defines structured units and quantities.
+"""
+
+# Standard library
+import operator
+
+import numpy as np
+
+from .core import Unit, UnitBase, UNITY
+
+
+__all__ = ['StructuredUnit']
+
+
+DTYPE_OBJECT = np.dtype('O')
+
+
+def _names_from_dtype(dtype):
+    """Recursively extract field names from a dtype."""
+    names = []
+    for name in dtype.names:
+        subdtype = dtype.fields[name][0]
+        if subdtype.names:
+            names.append([name, _names_from_dtype(subdtype)])
+        else:
+            names.append(name)
+    return tuple(names)
+
+
+def _normalize_names(names):
+    """Recursively normalize, inferring upper level names for unadorned tuples.
+
+    Generally, we want the field names to be organized like dtypes, as in
+    ``(['pv', ('p', 'v')], 't')``.  But we automatically infer upper
+    field names if the list is absent from items like ``(('p', 'v'), 't')``,
+    by concatenating the names inside the tuple.
+    """
+    result = []
+    for name in names:
+        if isinstance(name, str) and len(name) > 0:
+            result.append(name)
+        elif (isinstance(name, list)
+              and len(name) == 2
+              and isinstance(name[0], str) and len(name[0]) > 0
+              and isinstance(name[1], tuple) and len(name[1]) > 0):
+            result.append([name[0], _normalize_names(name[1])])
+        elif isinstance(name, tuple) and len(name) > 0:
+            new_tuple = _normalize_names(name)
+            result.append([''.join([(i[0] if isinstance(i, list) else i)
+                                    for i in new_tuple]), new_tuple])
+        else:
+            raise ValueError(f'invalid entry {name!r}. Should be a name, '
+                             'tuple of names, or 2-element list of the '
+                             'form [name, tuple of names].')
+
+    return tuple(result)
+
+
+class StructuredUnit:
+    """Container for units for a structured Quantity.
+
+    Parameters
+    ----------
+    units : unit-like, tuple of unit-like, or `~astropy.units.StructuredUnit`
+        Tuples can be nested.  If a `~astropy.units.StructuredUnit` is passed
+        in, it will be returned unchanged unless different names are requested.
+    names : tuple of str, tuple or list; `~numpy.dtype`; or `~astropy.units.StructuredUnit`, optional
+        Field names for the units, possibly nested. Can be inferred from a
+        structured `~numpy.dtype` or another `~astropy.units.StructuredUnit`.
+        For nested tuples, by default the name of the upper entry will be the
+        concatenation of the names of the lower levels.  One can pass in a
+        list with the upper-level name and a tuple of lower-level names to
+        avoid this.  For tuples, not all levels have to be given; for any level
+        not passed in, default field names of 'f0', 'f1', etc., will be used.
```

Comment:
I found this hard to follow without examples. Maybe this could be demonstrated in an Examples block, with demonstrations of each of these cases?

Reply:
OK, makes sense. I did notice I ended up rarely using `names`...